Smart url parsing

Trombach opened the pull request:
## Description

This PR resolves #2116 

Update: The behaviour is now a bit more similar to how Github handles this. I have also updated the video to show the new behaviour

Adds logic to detect if a paste event contains a URL. If true the URL will be formatted with the appropriate markdown. Open to suggestions on how to handle text selection after the paste event. The current implementation is demonstrated in the videos below.

It might also be better to implement this in the same way github does this, only adding formatting if text is already selected and otherwise just pasting the plain link, which might make things easier to read in the editor if no alternative link text is desired.

## Screenshots

### Before
[Screencast from 2023-09-24 17-30-03.webm](https://github.com/LemmyNet/lemmy-ui/assets/19306765/9b175520-77fa-42af-880c-fe16012c092f)

### After
[Screencast from 2023-09-26 21-59-39.webm](https://github.com/LemmyNet/lemmy-ui/assets/19306765/55bcdac4-c705-4301-adb6-c10a9487c266)

SleeplessOne1917 commented:
> It might also be better to implement this in the same way github does this, only adding formatting if text is already selected and otherwise just pasting the plain link, which might make things easier to read in the editor if no alternative link text is desired.

This is the better approach in my opinion. I'll wait for others to chime in on how it should be done before checking out the code.

dessalines commented:
I agree with sleepless, it'd be better to detect selected text first, then a paste event can turn it into a markdown link.

SleeplessOne1917 commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -374,10 +374,59 @@ export class MarkdownTextArea extends Component<
+      const selectedText = i.getSelectedText() || url.toString();
+
+      // update textarea content
+      i.setState({
+        content: `${
+          i.state.content ? i.state.content.substring(0, selectionStart) : ""
+        }[${selectedText}](${url.toString()})${
+          i.state.content ? i.state.content.substring(selectionEnd) : ""
+        }`,
+      });
```
I'm pretty sure inferno has similar issues to react when trying to set state using a previous value of that state. Try this:
```suggestion
     i.setState(({content}) => ({
         content: `${content?.substring(0, selectionStart) ?? ""}[${selectedText}](${url.toString()})${content?.substring(selectionEnd) ?? ""}`
     }));
```

You will probably need to format this with prettier still.

SleeplessOne1917 commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -374,10 +374,59 @@ export class MarkdownTextArea extends Component<
+    if (textarea instanceof HTMLTextAreaElement) {
+      event.preventDefault();
+      const { selectionStart, selectionEnd } = textarea;
+      const selectedText = i.getSelectedText() || url.toString();
```
If the return value of `getSelectedText` can be null or undefined, it is slightly clearer to use `??` instead of `||`.

Trombach replied:
I actually used `||` specifically since `getSelectedText` can't return `null` or `undefined` but could return `""`. I wanted to fallback to the url in that case, but since the consensus seems to be that the smart formatting should not be applied if no text is selected this fallback will become obsolete.

Trombach commented:
@SleeplessOne1917 @dessalines Thanks for the initial feedback. I have updated the PR and brought the behaviour closer to what Github does. I have updated the video as well

SleeplessOne1917 commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -374,10 +374,63 @@ export class MarkdownTextArea extends Component<
+    }
+  }
+
+  isValidUrl(value: string) {
```
This function should either be renamed or inlined. Seeing a url returned from what the name makes seem should be a boolean threw me through a loop for a second.

dessalines replied:
Also, make sure the return type is explicitly named: `: bool`

SleeplessOne1917 commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -374,10 +374,63 @@ export class MarkdownTextArea extends Component<
+    const url = i.isValidUrl(event.clipboardData.getData("text"));
+    if (url) {
+      i.handleUrlPaste(url, i, event);
+      return;
```
This return statement is unnecessary since it's at the end of the function anyway.

dessalines commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -374,10 +374,63 @@ export class MarkdownTextArea extends Component<
+      const { selectionStart, selectionEnd } = textarea;
+
+      // if no selection, just insert url
+      if (selectionStart === selectionEnd) return;
```
How is it doing anything here?

Trombach replied:
The comment explains it. If no text is selected (i.e. the curser is in one position and therefore `selectionStart === selectionEnd`) we don't want to apply formatting

dessalines replied:
Gotcha, it has to do with event.preventDefault() not being called.

dessalines commented on src/shared/components/common/markdown-textarea.tsx:
```diff
@@ -374,10 +374,63 @@ export class MarkdownTextArea extends Component<
+  }
+
+  isValidUrl(value: string) {
+    if (!value) return false;
```
This isn't necessary, you aren't passing in `string | undefined`, so we can assume its there.

Trombach replied:
Due to the beauty of javascript and `""` being falsy, it does something :) It actually prevents a constructor call to `URL` if you pass in an empty string which would never be a valid URL. I'm moving to the `validUrl` helper function, so it won't be relevant.

dessalines reviewed (commented):
I can't wait to move all this to rust so I can make everyone stop doing early returns ;)

IE `return if (condition) { } else if (condition) { } else { }`
